packages/react-sdk/src/components/FallbackSnapshotProvider/FallbackSnapshotProvider.tsx

```diff
@@ -0,0 +1,154 @@
+  const whiteboardInstance = useActiveWhiteboardInstance();
+  const { data: roomMembers } = useGetRoomMembersQuery();
+  const { data: roomEncryption } = useGetRoomEncryptionQuery();
+  const { data: roomHistoryVisibility } = useGetRoomHistoryVisibilityQuery();
+  const ownUserId = useWidgetApi().widgetParameters.userId;
+
+  useEffect(() => {
+    const isRoomEncrypted = checkIfRoomEncrypted(roomEncryption);
+    const roomHistoryVisibilityValue = getRoomHistoryVisibilityValue(
+      roomHistoryVisibility,
+    );
+    const isRoomHistoryVisibilityShared = checkIfRoomHistoryVisibilityShared(
+      roomHistoryVisibilityValue,
+    );
+    const sendFallbackSnapshot = shouldSendFallbackSnapshot(
+      isRoomEncrypted,
+      isRoomHistoryVisibilityShared,
+      roomHistoryVisibilityValue,
+    );
+
+    if (sendFallbackSnapshot) {
+      const membershipFilter = getMembershipFilter(roomHistoryVisibilityValue);
+      const members = roomMembers?.entities;
+
+      if (members) {
+        const invitesOrJoins = Object.values(members).filter(membershipFilter);
+        const sortedInvitesOrJoins = invitesOrJoins.sort((a, b) => {
+          return b.origin_server_ts - a.origin_server_ts;
+        });
+        const lastInviteOrJoin = sortedInvitesOrJoins[0];
+
+        if (
+          shouldUseInviteTimestamp(lastInviteOrJoin, roomHistoryVisibilityValue)
+        ) {
+          // TODO: Implement logic to get the timestamp of the invite event.
+          // There is currently no way to get an event by id or
+          // fetch an older state event from the room timeline using the Widget API
+        }
+
+        const immediate = shouldPersistImmediately(lastInviteOrJoin, ownUserId);
+
+        whiteboardInstance.persist({
+          timestamp: lastInviteOrJoin.origin_server_ts,
+          immediate,
+        });
+      }
+    }
+  }, [
+    roomMembers,
+    whiteboardInstance,
+    roomEncryption,
+    roomHistoryVisibility,
+    ownUserId,
+  ]);
+
+  return <>{children}</>;
+}
+
+function checkIfRoomEncrypted(roomEncryption?: {
+  event: StateEvent<RoomEncryptionEvent> | undefined;
+}): boolean {
+  return (
+    (roomEncryption && roomEncryption.event?.content?.algorithm) !== undefined
+  );
+}
+
+function getRoomHistoryVisibilityValue(roomHistoryVisibility?: {
+  event: StateEvent<RoomHistoryVisibilityEvent> | undefined;
+}): string | undefined {
+  return roomHistoryVisibility?.event?.content?.history_visibility;
+}
+
+function checkIfRoomHistoryVisibilityShared(
+  roomHistoryVisibilityValue: string | undefined,
+): boolean {
+  return roomHistoryVisibilityValue === 'shared';
+}
+
+function shouldSendFallbackSnapshot(
+  isRoomEncrypted: boolean,
+  isRoomHistoryVisibilityShared: boolean,
+  roomHistoryVisibilityValue: string | undefined,
+): boolean {
+  return (
+    isRoomEncrypted ||
+    (!isRoomHistoryVisibilityShared && roomHistoryVisibilityValue !== undefined)
```

Comment:
Why is the value checked here? Inside `checkIfRoomHistoryVisibilityShared` it is already tested to be `shared`. I would move everything into `checkIfRoomHistoryVisibilityShared`.